### PR TITLE
{lib}[foss/2021b] CuPy v11.4.0 w/ Python 3.9.6

### PR DESCRIPTION
**Pull request by Micket**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by Micket** [comment]
So.. I opted to not update the patch that was used previously. 
It's possible to skip sanity check steps, and, if that patch it used when doing so one would have a broken installation, with leftover code that doesn't belong there.

It also felt a bit fragile to rely on that. I think it's just better for those who really need to install this via nodes without installed CUDA should just skip the sanity checks.

**Comment by Micket** [comment]
@boegelbot please test @ generoso

**Comment by Micket** [comment]
Test report by @Micket
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
alvis-c1 - Linux Rocky Linux 8.5, x86_64, Intel Xeon Processor (Skylake), Python 3.6.8
See https://gist.github.com/e297b3938b8f18610645db7f3f9789df for a full test report.

**Comment by Micket** [comment]
eh, ok, i guess generoso also skips the cuda drivers. :/

**Comment by boegel** [comment]
> So.. I opted to not update the patch that was used previously. It's possible to skip sanity check steps, and, if that patch it used when doing so one would have a broken installation, with leftover code that doesn't belong there.
> 
> It also felt a bit fragile to rely on that. I think it's just better for those who really need to install this via nodes without installed CUDA should just skip the sanity checks.

What's currently done in the patch can easily be done in `postinstallcmds`, no?
Maybe still worth doing like this?

**Comment by Micket** [comment]
I believe it was hackier than you think.

Previous easyconfig had postinstallcmds that copied extra stubs into the installation, in order to allow systems without CUDA to attempt to import the library without errors during sanity check

But we don't want these libraries present afterwards, so the patch hooks in extra commands in the installed __init__.py file, which removes these additional libraries and `sed`-comments itself.
So, the act of `import cupy` modified the installation at that point, cleaning up this hack and removing itself.

Basically making a secret "postsanitycheckcmds"

All to trick the node into accepting sanity check command (which will immediately break after the first use if the node doesn't have cuda drivers as a backup).

**Comment by boegel** [comment]
@Micket As discussed in today's conf call (see notes [here](https://github.com/easybuilders/easybuild/wiki/Conference-call-notes-20230118#qa)), let's get rid of that ugly hack: it's not OK that the sanity check triggers changes to the installation by injecting self-modifying code that triggers `rm` commands...

**Comment by Micket** [comment]
Test report by @Micket
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
vera-r07-03 - Linux Rocky Linux 8.5, x86_64, Intel(R) Xeon(R) Gold 6338 CPU @ 2.00GHz, 4 x NVIDIA NVIDIA A40, 510.73.08, Python 3.6.8
See https://gist.github.com/da7a97ce93f169e1bcb78212c106c701 for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
vera41-1 - Linux Rocky Linux 8.5, x86_64, Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz, 1 x NVIDIA Tesla T4, 510.73.08, Python 3.6.8
See https://gist.github.com/a9f0715538613d12a466e4c0060cf338 for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2872
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
vera-r07-03 - Linux Rocky Linux 8.5, x86_64, Intel(R) Xeon(R) Gold 6338 CPU @ 2.00GHz, 4 x NVIDIA NVIDIA A40, 510.73.08, Python 3.6.8
See https://gist.github.com/f7f3d10df7c1bc86323e32b3b6e797c0 for a full test report.

File: easybuild/easyconfigs/c/CuPy/CuPy-11.4.0-foss-2021b-CUDA-11.4.1.eb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    ('hypothesis', '6.54.6', {
+        'checksums': ['2d5e2d5ccd0efce4e0968a6164f4e4853f808e33f4d91490c975c98beec0c7c3'],
+    }),
+    ('fastrlock', '0.5', {
+        'checksums': ['9ae1a31f6e069b5f0f28ba63c594d0c952065de0a375f7b491d21ebaccc5166f'],
+    }),
+    ('cupy', version, {
+        'module': False,
+        'runtest': "cd tests && pytest",
+        'testinstall': True,
```
Comment (akesandgren) [inline easybuild/easyconfigs/c/CuPy/CuPy-11.4.0-foss-2021b-CUDA-11.4.1.eb:35]: testinstall is not picked up here due to being used in an extension. We need a fix in framework I guess

Comment (Micket) [reply easybuild/easyconfigs/c/CuPy/CuPy-11.4.0-foss-2021b-CUDA-11.4.1.eb:35]: https://github.com/easybuilders/easybuild-easyblocks/pull/2872 fixed here

Comment (akesandgren) [reply easybuild/easyconfigs/c/CuPy/CuPy-11.4.0-foss-2021b-CUDA-11.4.1.eb:35]: No, it doesn't because framework doen't progress testinstall up to pythonpackage.py when it's in an extension.

Comment (akesandgren) [reply easybuild/easyconfigs/c/CuPy/CuPy-11.4.0-foss-2021b-CUDA-11.4.1.eb:35]: meh, managed to pick the wrong version of that easyblock :-(

File: easybuild/easyconfigs/c/CuPy/CuPy-11.4.0-foss-2021b-CUDA-11.4.1.eb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    ('NCCL', '2.10.3', versionsuffix),
+    ('cuTENSOR', '1.6.1.5', versionsuffix, SYSTEM),
+]
+
+use_pip = True
+
+exts_default_options = {'source_urls': [PYPI_LOWER_SOURCE]}
+
+exts_list = [
+    ('hypothesis', '6.54.6', {
```
Comment (Micket) [inline easybuild/easyconfigs/c/CuPy/CuPy-11.4.0-foss-2021b-CUDA-11.4.1.eb:26]: I have missed updated this PR with this, but i also gave up on hypothesis and pytest versions that cupy prefers; those newer versions of pytest and hypothesis requires python 3.11. 

Instead, needs to exclude the single test uses hypothesis instead. 
But i still dont know what to do with cupyx tests that crash hard.

**Comment by akesandgren** [comment]
For starters see https://github.com/Micket/easybuild-easyconfigs/pull/15

**Comment by akesandgren** [comment]
CuPy tests have some quirks.
cupyx_tests/distributed_tests fails (for me) all NCCL tests without MPI backend with
```
cupy_backends.cuda.api.runtime.CUDARuntimeError: cudaErrorInitializationError: initialization error
```
That test is also explicitly calling mpiexec -n 2 for the NCCL with MPI backend tests and will fail if run inside a srun/mpirun context, i.e. if pytest is started as "mpirun -n 1 pytest".

If one doesn't set CUPY_TEST_GPU_LIMIT to the number of actually available GPUs during the test it will fail the tests marked multi_gpu(x) where x > the actual number of available GPUs.

Some tests in cupy_tests/testing_tests fail for, currently, unknown reasons (haven't investigated yet).

cupyx_tests/scipy_tests have ~150 failed tests, no clue as to why yet.

**Comment by akesandgren** [comment]
The scipy tests differs numerically from the expected values for a bunch of tests. Not going to investigate that.

We can either avoid running the test altogether or filter ouyt the problematic ones.

**Comment by akesandgren** [comment]
```
'runtest': 'pytest --ignore=tests/cupyx_tests/scipy_tests --ignore=tests/cupyx_tests/distributed_tests --ignore=tests/cupyx_tests/tools_tests --ignore=tests/example_tests --ignore=tests/cupy_tests/testing_tests/test_parameterized.py tests
```
should get a lot of testing done without failing and the scipy tests can be micromanaged if we want that.

**Comment by akesandgren** [comment]
Superceeded by #17526
